Parameter grouping

Exilliar opened the pull request:
Allows for users, on the assessment page, to group their objects for material selection by any _string_ property on their model. Only string properties to avoid grouping by things like volume, which wouldn't make sense (this is also the way that Arup Carbon handled this feature). I haven't done any other filtering of the parameters though, so it is possible to group by things like "element id" (every object has a unique element id), which would probably be bad, but I don't know which parameters should be excluded from here so I've just left it to the user to decide what they want to select. I think we can work on that in the future. It does slow down the loading of the assessment page a bit, but I think this is worth it.

The parameter's are also filtered so that only the params that are present on evey object to avoid having objects without a group (although it would be possible to just throw the objects without the param into an "undefined" group, or something similar)

Some parameter values are longer than the pill's that they're in (like the below pic). I think that this is an issue for a later pr? Pretty sure this was an issue that was present before this pr.
![overflow](https://user-images.githubusercontent.com/43271984/222515200-1d231aa5-0727-4b8f-8b2c-b40d3e47cb97.PNG)

This is also not the prettiest code that I've ever written, but I mainly just wanted to get this done as it's been heavily requested.

Some parts borrow (ie, copy/pasted) fairly heavily from the Speckle codebase ([I'm pretty sure that the Speckle licence allows for this](https://github.com/specklesystems/speckle-server/blob/main/LICENSE)) and I haven't really spent too much time making it look nice for the same reason as above.

I also have not yet tested this on any Rhino models (I've done a quick test with the old one that we have, and it didn't work, but I think that that model might just generally be broken)

To test simply run a report and see what you can group by any param.

ro-sa commented on src/views/Assessment.vue:
```diff
@@ -240,6 +248,13 @@ export default class Assessment extends Vue {
+  materialGroupSelected(objectGroup: string) {
+    this.types = this.findTypes(this.groupingProps, objectGroup);
+    this.selectedObjectGroup = objectGroup;
+    // not sure if the materials actually get reset when the grouping changes, but safest to just say that it does
```
agreed. please remove comment.

ro-sa reviewed (commented):
please check that assigning carbon factors to a custom group won't be overwritten if the same objects get assigned carbon factors as part of a `MaterialGroup`.

Exilliar commented:
> please check that assigning carbon factors to a custom group won't be overwritten if the same objects get assigned carbon factors as part of a `MaterialGroup`.

Okay, so did some testing. Looks like if you make a custom group, then if you don't change what parameter is being grouped then everything works like you'd expect.

If you change what parameter you're grouping by after making a custom group, then that group will dissapear. Technically, if you assigned any materials to the objects in that custom group, then those objects will still have that material value. But when you then select a new material from the group that those objects are now part of the old material value will be over-written. So should be fine.

There was a bug where if you:
1. made a custom group
2. changed object grouping
3. tried to make another custom group with the same name as the old custom group

then the creation of that group would fail as it would count as a duplicate. I've fixed that bug and pushed up a commit to fix.

Exilliar commented on src/views/utils/propertyFiltering/findStringProps.ts:
```diff
@@ -0,0 +1,129 @@
+import { flattenObject } from "./";
+import { ObjectsObj, StringPropertyGroups } from "@/models/newAssessment";
+
+// PRETTY MUCH ALL OF THIS CODE (AND SOME OF THE COMMENTS) IS FROM THE SPECKLE VIEWER/FRONTEND, REPO HERE: https://github.com/specklesystems/speckle-server
```
changed to `source: https://github.com/specklesystems/speckle-server` just to make sure that we know where the code is from incase something goes wrong with it in the future
